### DISCUSSION TITLE
how to use contains / containsany functions in filter ?

### DISCUSSION THREAD

**Discussion by kran** [ArcadeData/arcadedb#482] [Q&A]
I have field named `stages` like this: 

```
[
  {
    "t": "10262",
    "d": 16593,
    "u": "u1",
    "fs": "A",
    "f": "10262",
    "ts": "A"
  },
  {
    "t": "10622",
    "d": 16593,
    "u": "u2",
    "fs": "A",
    "f": "10262",
    "ts": "B"
  },
  {
    "t": "10001",
    "d": 16593,
    "u": "u2",
    "fs": "B",
    "f": "10622",
    "ts": "C"
  }
]
```

then I want to fetch records stages have `t` = 10001, query wrote like this, but not work, always return empty result: 

```
select from SomeType where stages contains(t = '10001')
```

then rewrote to, and this time got the matched records:

```
select from SomeType where stages.t contains '10001'
```

I want to know how `contains*` functions work

**Comment by kran** [comment]
and other question is how to remove a column from types ?

**Reply by lvca** [reply]
You can use the `DROP PROPERTY` SQL command or Java API.

**Reply by kran** [reply]
`drop property` does not work, after drop property,  column with null value just returned. 
I want to  drop column completely (something like mysql).

**Comment by lvca** [comment]
The `contains` operator is supposed to work exactly as you described. I wrote a test case and works:

```java

  @Test
  public void testContains() {
    String className = "testContains";

    DocumentType clazz1 = database.getSchema().createDocumentType(className);
    clazz1.createProperty("list", Type.LIST);

    database.getSchema().createDocumentType("embeddedList");

    database.transaction(() -> {
      for (int i = 0; i < 100; i++) {
        MutableDocument document = database.newDocument(className);
        document.set("list", new ArrayList<>());

        for (int j = i; j < i + 3; j++)
          document.newEmbeddedDocument("embeddedList", "list").set("value", j);

        document.save();
      }
    });

    int totalFound = 0;
    for (ResultSet result = database.query("sql", "select from " + className + " where list contains ( value = 3 )"); result.hasNext(); ) {
      Result item = result.next();
      List<EmbeddedDocument> embeddedList = item.getProperty("list");

      List<Integer> valueMatches = new ArrayList<>();
      for (EmbeddedDocument d : embeddedList)
        valueMatches.add(d.getInteger("value"));

      Assertions.assertTrue(valueMatches.contains(3));

      ++totalFound;
    }

    Assertions.assertEquals(3, totalFound);
  }
```

**Reply by kran** [reply]
thanks @lvca 
it's wired, your test code returns the 3 records. 
after then I confirmed many times, the sql `field contains (key = val)` return nothing for my type, but `field.key contains val` just works. 
maybe the only different is that I insert records with `update SomeType merge $JSON upsert where xx`?
`select stages.type() from SomeType` returns `LIST` that I think it's the right property type. 

any suggestion?

**Reply by kran** [reply]
I notice that the test code creates embedded document, and after serialize it's json string value is :

```
[
  {
    "@type": "embeddedList",
    "@cat": "d",
    "value": 0
  },
  {
    "@type": "embeddedList",
    "@cat": "d",
    "value": 1
  },
  {
    "@type": "embeddedList",
    "@cat": "d",
    "value": 2
  }
]
```

but the way insert record with `update merge $JSON upsert`, property has just original keys(no `@cat`, `@type`)

**Reply by lvca** [reply]
Are you using a map as an embedded type?

**Reply by kran** [reply]
no, it's list, the entire document is like this: 

```
{
  "@rid": "#1:1",
  "stages":  [
      {
        "t": "10262",
        "d": 16593,
        "u": "u1",
        "fs": "A",
        "f": "10262",
        "ts": "A"
      },
    .....
  ],
  "others": ....

```

**Reply by lvca** [reply]
No metadata (@cat and @type) so seems like a map to me. I've changed the test case and actually with embedded maps I was able to reproduce your problem. I was able to fix this case with a map. Could you please get the latest snapshot (from git), compile it, and test it again? Thanks in advance.

**Reply by lvca** [reply]
(Commit 5e7dfb7a)

**Reply by kran** [reply]
thanks @lvca.

i'll try it in this weekend.

but here i have a question, why it's a map?

`stages.type()` gives `list` as the answer.

**Reply by lvca** [reply]
Yes, it's a list of maps instead of a list of embedded documents. Maps and embedded documents are very similar. The reason must be in using the `merge` from a json. The JSON is converted in map(s).

**Reply by kran** [reply]
ok, got it, i will try again.

luca, i like the idea of arcadedb, like the project  a lot. and using it in company products. but with the experience of these weeks, i have to ask myself if it's a right choice. i can bypass bugs in sql layer, but then i'm a little worried about the stablity of storage engine layer. 

i always know that it is my responsibility not yours. but i just worry about it.

forgive me if i'm not so friendly. 

thank you.

**Reply by lvca** [reply]
@kran that's a legitimate question! :-) The ArcadeDB engine is 100X more stable than OrientDB which has a much longer history. Unfortunately, we forked the entire SQL engine from OrientDB inheriting all its bugs. While nobody is fixing those bugs in OrientDB anymore, we're active on ArcadeDB. I have to say that the most stable part of ArcadeDB is the engine, and other layers (like the Postgres protocol) are not very mature and there is still a lot of work.
